### ISSUE TITLE
WebGL context corruption when running multiple overlaid Deck instances in tests

### ISSUE THREAD

**Issue by chrisgervang** [visgl/deck.gl#9950]
@ibgreen @felixpalmer @Pessimistress I ran into a strange issue where adding more than one test to a PR caused CI to hang in https://github.com/visgl/deck.gl/pull/9947/changes/db98854595b08c806a6352cbe6441e4ff9c4411a, specifically the new overlaid tests. It didn't matter which of the 4 I added, 2 or more and the tests later hang on `onAfterRender` at the same spot.

I'm opening this in case others have seen a similar issue.

Here is an AI investigation into the causes. I'm not convinced any one of these are the root cause yet.. curious if any of these sound like real problems to you.

## Problem

When running multiple tests that create overlaid Deck instances (e.g., `MapboxOverlay` in overlaid mode), the shared WebGL context accumulates leftover state that prevents subsequent tests from initializing properly, causing them to hang indefinitely when they try to create new Deck instances with `onAfterRender` callbacks.

### Symptoms
- A single overlaid test works perfectly ✓
- Multiple overlaid tests in sequence cause subsequent tests in other files to hang ✗ (see https://github.com/visgl/deck.gl/pull/9947/changes/db98854595b08c806a6352cbe6441e4ff9c4411a)
- Tests hang specifically when creating a new Deck with `onAfterRender` - the callback never fires
- Interleaved mode tests work fine with multiple tests

### Root Cause

The investigation revealed multiple issues with how Deck/luma.gl handle WebGL context cleanup, resulting in **state pollution** of the shared gl context:

#### 1. **Event Listeners Never Removed**
When luma.gl creates a `WebGLDevice` from an existing gl context, it attaches event listeners to the canvas:

```javascript
canvas.addEventListener('webglcontextlost', onContextLost, false);
canvas.addEventListener('webglcontextrestored', onContextRestored, false);
```

**Problem**: These listeners are **never removed** when the Device is finalized or destroyed. These orphaned listeners persist and can interfere with subsequent WebGL operations.

**Location**: `@luma.gl/webgl/dist/context/helpers/create-browser-context.js`

#### 2. **Stale References Attached to Shared gl Context**
When `WebGLDevice` attaches to a gl context, it sets properties that persist after finalization:

```javascript
this.gl.device = this;      // Link WebGL context back to device
this.gl._version = 2;       // Store WebGL version
```

**Problem**: When `Deck.finalize()` is called, these properties remain on the shared gl object:
- `gl.device` - points to a finalized device instance (stale reference)
- `gl._version` - persists unnecessarily
- `gl._extensions` - object references remain
- WebGLStateTracker state not fully reset

When a new Deck tries to initialize with this polluted context, it detects these existing references and may fail to initialize cleanly.

**Location**: `@luma.gl/webgl/dist/adapter/webgl-device.js` (lines 84, 118, 134)

#### 3. **Shared Singleton Context in Tests**
The test-utils module uses a **global singleton** WebGL context:

```typescript
export const gl = webglDevice?.gl || 1;
globalThis.glContext = globalThis.glContext || gl;
```

**Problem**: All tests share the same gl context. Each overlaid Deck test adds more leftover state. By the time subsequent test files run, the accumulated state pollution prevents new Deck instances from initializing properly.

**Location**: `modules/test-utils/src/utils/setup-gl.ts`

#### 4. **Why Interleaved Mode Works**
Interleaved mode extracts the gl context from mapbox (`map.painter.context.gl`) rather than creating its own Device. Mapbox properly manages the context lifecycle, so multiple interleaved tests reuse the same gl context cleanly.

### What's Actually Happening

The WebGL context itself isn't corrupted or damaged. Rather:
- **Incomplete cleanup**: Properties and listeners aren't properly removed during finalization
- **State pollution**: The shared gl context accumulates leftover state from each test
- **Resource leakage**: Event listeners and object references aren't released
- **Stale references**: Properties point to finalized/destroyed objects

This accumulated baggage prevents new Deck instances from initializing cleanly, causing the animation loop to fail and `onAfterRender` callbacks to never fire.

## Steps to Reproduce

```typescript
import test from 'tape-promise/tape';
import {Deck} from '@deck.gl/core';
import {MapboxOverlay} from '@deck.gl/mapbox';
import {ScatterplotLayer} from '@deck.gl/layers';
import {gl} from '@deck.gl/test-utils';

// First overlaid test - works fine
test('First overlaid test', t => {
  const map = new MockMapboxMap();
  const overlay = new MapboxOverlay({
    layers: [new ScatterplotLayer({id: 'test1'})]
  });
  map.addControl(overlay);
  map.removeControl(overlay);
  t.end();
});

// Second overlaid test - works fine in isolation
test('Second overlaid test', t => {
  const map = new MockMapboxMap();
  const overlay = new MapboxOverlay({
    layers: [new ScatterplotLayer({id: 'test2'})]
  });
  map.addControl(overlay);
  map.removeControl(overlay);
  t.end();
});

// Subsequent test in a different file - HANGS
test('Deck instance after multiple overlaid tests', t => {
  const deck = new Deck({
    gl,
    layers: [new ScatterplotLayer({id: 'test3'})],
    onAfterRender: () => {
      // This callback NEVER fires - test hangs forever
      t.ok(deck, 'Deck rendered');
      deck.finalize();
      t.end();
    }
  });
});
```

## Impact

This issue makes it impossible to:
- Write comprehensive test suites for overlaid MapboxOverlay behavior
- Test multiple overlaid scenarios in the same test run
- Run tests with proper isolation when using the shared gl context from test-utils

### Solution 1: Clean Up gl Context State in Deck.finalize() (Quick Fix)

Add cleanup logic to remove attached properties from the gl context:

```typescript
finalize() {
  // Existing cleanup...
  this.animationLoop?.stop();
  this.animationLoop?.destroy();
  
  // NEW: Clean up gl context state to prevent pollution
  if (this.device?.gl) {
    delete this.device.gl.device;
    delete this.device.gl._version;
    delete this.device.gl._extensions;
  }
  
  // Existing cleanup continues...
}
```

### Solution 2: Fix luma.gl Device to Remove Event Listeners (Proper Fix)

Update `@luma.gl/webgl` to properly clean up event listeners when Device is destroyed:

```javascript
// In WebGLDevice.destroy()
destroy() {
  if (this.canvas) {
    this.canvas.removeEventListener('webglcontextlost', this._onContextLost);
    this.canvas.removeEventListener('webglcontextrestored', this._onContextRestored);
  }
  // Existing cleanup...
}
```

This requires storing references to the bound event handler functions.

### Solution 3: Reset Shared gl Context Between Tests (Test Infrastructure)

Add a utility function to reset the shared gl context to a clean state:

```typescript
// In test-utils
export function resetGLContext() {
  if (gl.device) {
    delete gl.device;
    delete gl._version;
    delete gl._extensions;
  }
  // Reset any WebGLStateTracker state
  // Clear any lingering event listeners
}

// In tests
afterEach(() => {
  resetGLContext();
});
```

### Solution 4: Use Fresh gl Context Per Test (Test Isolation)

Instead of sharing a global singleton, create a fresh gl context for each test that needs it:

```typescript
// In test-utils
export function createTestGLContext() {
  const canvas = document.createElement('canvas');
  return canvas.getContext('webgl2');
}

// In tests
test('My test', t => {
  const gl = createTestGLContext();
  const deck = new Deck({gl, ...});
  // Test runs with isolated context
});
```

## Recommended Approach

I recommend implementing **Solutions 1, 2, and 3 together**:

1. **Short term**: Add cleanup in `Deck.finalize()` (Solution 1) and `resetGLContext()` utility (Solution 3) to unblock test writing
2. **Long term**: Fix luma.gl Device to properly clean up event listeners (Solution 2) as the proper architectural fix
3. **Best practice**: Document that tests using shared gl contexts should call `resetGLContext()` between tests

## Additional Context

This issue was discovered while adding tests for automatic view injection in MapboxOverlay. The interleaved mode tests work fine because they use mapbox's managed gl context, but overlaid mode tests pollute the shared test-utils gl context.

### Related Files
- `modules/core/src/lib/deck.ts` - Deck.finalize()
- `modules/test-utils/src/utils/setup-gl.ts` - Shared gl context
- `@luma.gl/webgl/dist/adapter/webgl-device.js` - Device initialization
- `@luma.gl/webgl/dist/context/helpers/create-browser-context.js` - Event listener attachment

**Comment by ibgreen** [comment]
Thanks for putting this together.
- I am not quite sure which proposed changes would definitely fix the tests and which changes are more observations of things that potentially could cause confusion/issues.
- But I agree that the code that attaches data to the WebGL context could use some cleanup to minimize its footprint and eliminate potential "red herrings" for maintainers and LLMs.
- I am putting up a series of PRs to consolidate all context injected data in the `gl.luma` object and centralize updates and improve typing.
- Also considering adding a `Device.detach()` method that unregisters the context lost/restored handlers.
- This low level stuff can easily get brittle, so taking a bottoms up approach is likely to pay off, even if it takes a bit more time to get to a fix.

Notes:
- As to obtaining WebGL contexts in test, there are methods for context creation/reuse in @luma.gl/test-utils that we may want to adopt more generally in deck.gl and then we can refine context handling in one place. These methods also allow for WebGPUDevice testing.
- Traditionally WebGL contexts start to get destroyed by the browser once a limit is reached (16 or 32) so this could also cause flaky tests and is a big reason why we don't mint fresh contexts for every test.

**Comment by chrisgervang** [comment]
Thanks for looking into these!

> As to obtaining WebGL contexts in test, there are methods for context creation/reuse in @luma.gl/test-utils that we may want to adopt more generally in deck.gl and then we can refine context handling in one place. These methods also allow for WebGPUDevice testing.

That sounds like a good direction. I'm modernizing deck's testing infra to vitest and we could consider how to use these

> Traditionally WebGL contexts start to get destroyed by the browser once a limit is reached (16 or 32) so this could also cause flaky tests and is a big reason why we don't mint fresh contexts for every test.

Deck is warning of destroyed contexts in every CI test run - something to dig more into

**Comment by ibgreen** [comment]
> I'm modernizing deck's testing infra to vitest

That's awesome. If you haven't already, check out deck.gl-community which has already done that migration. 

There were a couple of tricky points I had to work through. 

And if we align we are better positioned to punch this upstream into dev-tools.

**Comment by chrisgervang** [comment]
Update: Creating a fresh device to use in the mapbox-overlay tests in overlaid mode fixes the hanging issue. See https://github.com/visgl/deck.gl/commit/61466c357f87a65302068018aa1908330a69a969

**Comment by ibgreen** [comment]
I believe that I have addressed the reported root causes in luma.gl 9.3. Will close this for now, if issue persists lets create a new ticket based on the current state.
